scripts/202306xx_populate_paymenttype_default/add_fixed_payment_type_default.js

```diff
@@ -4,31 +4,87 @@
 Add PaymentType.Fixed to all forms with existing payments_field object
 */
 
-// Count number of forms that contains payments_field property
-db.getCollection('forms')
-  .find({ payments_field: { $exists: true } })
-  .count()
+// Batch 1 2022
```

Comment:
Since all old forms can also be added with payments, it's best to scan the whole collection without causing performance issue. To do so, we need to utilise index which is `_id`, and we need to do a batch read/write. I write a script which you can refer to.

```
let startDate = toObjectId(new Date('2021-12-31T23:59:59.999+0800'))
const endDate = toObjectId(new Date())
const batchSize = 1000
let toUpdates = []
let totalUpdates = 0

function query(startDate, endDate) {
  return db
    .getCollection('forms')
    .find({
      payments_field: { $exists: true },
      'payments_field.payment_type': { $ne: 'Fixed' },
      _id: {
        $gt: startDate,
        $lte: endDate,
      },
    })
    .projection({ _id: 1 })
    .sort({ _id: 1 })
    .limit(batchSize)
}

function update() {
  if (toUpdates.length === 0) {
    console.log('nothing to update')
    return
  }
  const ids = toUpdates.map(function (doc) {
    return doc._id
  })
  const updateRes = db.getCollection('forms').updateMany(
    {
      _id: { $in: ids },
      payments_field: { $exists: true },
      'payments_field.payment_type': { $ne: 'Fixed' },
    },
    {
      $set: {
        'payments_field.payment_type': 'Fixed',
      },
    },
  )
  toUpdates = []
  console.log('updated', updateRes.modifiedCount)
}

while (startDate < endDate) {
  const res = query(startDate, endDate).toArray()
  if (res.length === 0) {
    break
  }
  toUpdates = toUpdates.concat(res)
  totalUpdates += res.length
  console.log(
    'Total toUpdate so far is',
    totalUpdates,
    'with startDate=',
    startDate,
    'and endDate=',
    endDate,
  )
  if (toUpdates.length >= batchSize) {
    update()
  }
  startDate = res[res.length - 1]._id
}

console.log("checking if there're anything else to update")
update()

function toObjectId(date) {
  var timestamp = Math.floor(date.getTime() / 1000)
  return ObjectId(timestamp.toString(16) + '0000000000000000')
}
```

Reply:
@mayying, was wondering if we should just ship the PR first as it's blocking other stuffs.

i think @KenLSM's current script works and it shouldn't take too much time as tested on the analytics node. Was thinking whether we push ahead and KIV this DB migration discussion for next time? 😅

(not sure if searching dates on `_id` work + @KenLSM made a good point that we are looking for `lastModified` and not `createdAt / _id`)

Reply:
Requested @foochifa to go ahead first, as this script is technically not required to be executed before we release. The code itself is backwards compatible. But it should be executed anytime before we release `By Products` to mitigate potential issues.